js/src/vm/JSScript.cpp

```diff
@@ -95,6 +95,186 @@ using JS::CompileOptions;
 using JS::ReadOnlyCompileOptions;
 using JS::SourceText;
 
+class FrontendAccessor {
```

Comment:
OK, you're making `JSScript::initFullyFrom` a template, for polymorphism (we need it to support input from two different sources). I think we should try a different approach to get the polymorphism we need.

- Rename `class FrontendAccessor` to `struct ScriptStencil`.

- Change most of the methods into fields:

    ```
    struct ScriptStencil {
      unsigned lineno;
      unsigned column;
      uint64_t nslots;
      bool strict;
      ...
      mozilla::Span<jsbytecode> code;
      mozilla::Span<jssrcnote> notes;
      ...
    };
    ```

- I think there will be a few remaining methods, like `finishInnerFunctions`. Make them pure virtual.

- Add a subclass that contains a private BytecodeEmitter* and implements the methods

- The subclass should have an `init` method that populates the fields and returns a success `bool` (since some of the code, like determining `nslots`, can fail).

- Change the type of the last argument of `JSScript::initFullyFrom` from `T*` to `const ScriptStencil&`.

And so on.

Reply:
2 questions:

 1. do we prefer `struct` for this case?  I thought we're going to use `class` for everything newly added
 2. isn't `virtual` call a performance problem?

Reply:
1. Oh, you're right. `class` will be fine. (I believe the style guide says it's OK to use `struct` for a plain object with just public data and no methods, but that's not what this is, for now.)

2. I don't think virtual method calls are *that* bad. How many remain? Three? One? If it regresses perf, I think we could fix it by switching `JSScript::initFullyFrom` to a template and putting `final` on the derived classes. But I'd be surprised if this was a problem.

Reply:
still implementing, but there are 6 virtual calls for now.
yes, I'll use virtual for now.